src/libcrun/kontain_nonkmexec.c

```diff
@@ -0,0 +1,347 @@
+   SLIST_HEAD_INITIALIZER(&execpath_state.execpath_head),
+   KONTAIN_OK_EXEC_TARGETS,
+   { 0, 0 }
+};
+
+static void
+kontain_execpath_freedb(execpath_state_t* statep)
+{
+   execpath_entry_t* epp;
+
+   while (!SLIST_EMPTY(&statep->execpath_head)) {
+      epp = SLIST_FIRST(&statep->execpath_head);
+      SLIST_REMOVE_HEAD(&statep->execpath_head, link);
+      free(epp->path);
+      regfree(&epp->re);
+      free(epp);
+   }
+   memset(&statep->execpath_mtime, 0, sizeof(statep->execpath_mtime));
+}
+
+// Destroy the execpath ok db.
+void
+libcrun_kontain_nonkmexec_clean(void)
+{
+   kontain_execpath_freedb(&execpath_state);
+}
+
+static void
+hash_2_ascii(BYTE hash[SHA256_BLOCK_SIZE], char *ascii)
+{
+   int i;
+   for (i = 0; i < SHA256_BLOCK_SIZE; i++) {
+      snprintf(&ascii[i * 2], 3, "%02x", hash[i]);
+   }
+   ascii[i * 2] = 0;
+}
+
+/*
+ * Compute the sha256 hash of the passed file.
+ * Return the hash in ascii in returned_hash.
+ * Returns:
+ *   0 - success
+ *   != 0 - an errno value describing what failed
+ */
+static int
+sha256_file(char* file, char *returned_hash)
+{
+   int f;
+   unsigned char filebuf[128*1024];
+   SHA256_CTX ctx;
+   BYTE final_hash[SHA256_BLOCK_SIZE];
+   ssize_t bytesread;
+
+   f = open(file, O_RDONLY);
+   if (f < 0) {
+      return errno;
+   }
+
+   sha256_init(&ctx);
+   while ((bytesread = read(f, filebuf, sizeof(filebuf))) > 0) {
+      sha256_update(&ctx, filebuf, bytesread);
+   }
+   if (bytesread < 0) {
+      return errno;
+   }
+   sha256_final(&ctx, final_hash);
+
+   close(f);
+
+   hash_2_ascii(final_hash, returned_hash);
+
+   return 0;
+}
+
+// Split a line from the exec ok file into its fields
+static int
+split_into_fields(char* dbentry, char *fields[])
+{
+   char* saveptr = NULL;
+
+   fields[0] = strtok_r(dbentry, ":", &saveptr);
+   if (fields[0] == NULL) {
+      return EINVAL;
+   }
+   fields[1] = strtok_r(NULL, ":", &saveptr);
+   if (fields[0] == NULL) {
+      return EINVAL;
+   }
+   fields[2] = strtok_r(NULL, ":", &saveptr);
+   if (fields[2] == NULL) {
+      return EINVAL;
+   }
+   return 0;
+}
+
+/*
+ * Read a execpath database and place it into execpath_state_t pointed to by statep.
+ * Entries in the file are a single line with fields separated by colons.
+ * There are 3 fields:
+ *   regular expression to match an input string against
+ *   the file to exec to if the re matches
+ *   a sha of the file being exec'ed to to verify the file is unchanged since the
+ *   db was created.
+ * Failures is this function cause the process to exit.
+ *
+ * We should probably use some sort of json representation for this instead of the
+ * archaic colon separated fields.
+ */
+static void
+kontain_execpath_builddb(execpath_state_t* statep)
+{
+   char buf[1024];
+   FILE* f = NULL;
+   char *fields[3];
+   execpath_entry_t* epp = NULL;
+   execpath_entry_t* tail = NULL;
+   int rc = 0;
+
+   if (statep->execpath_dbpath != NULL) {
+      kontain_execpath_freedb(statep);
```

Comment:
Why do we need to clean up the list ? And, how it can be not empty here ?

Reply:
The check for a non-null path doesn't need to be there.
Since this function is called because the caller wants the database to be re-read or initially read, we need to get rid of the old data if there is any before we create a new batch of data.

Reply:
well, that's exactly my question. krun is invoked once on each exec(). So this cannot be pre-filled, can't it ? So it looks like assert to me

Reply:
You are correct it is invoked once.  So, in theory there should be no need to destroy a pre-existing instance.
And, the execpath_dbpath is initialized statically and shouldn't change so there is no need to check for it being null.
I'd like to leave the code to destroy the preexisting database.  It does not add much overhead.
And, I have no real reason to keep it other than I never know how functions will be used in the future and
when I write a create function, I mostly write a destroy function too.